shipwright: 8.0.6 -> 9.0.2, update darwin dependencies and use applyPatches, include git build details, fix dev builds

qubitnano opened the pull request:
Closes https://github.com/NixOS/nixpkgs/issues/395651

https://github.com/HarbourMasters/Shipwright/releases/tag/9.0.0
https://github.com/HarbourMasters/Shipwright/releases/tag/9.0.1
https://github.com/HarbourMasters/Shipwright/releases/tag/9.0.2

~Untested on Darwin~



<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

qubitnano committed:
shipwright: 8.0.6 -> 9.0.2

matteo-pacini commented:
@qubitnano Thanks for this, will send out a patch for Darwin shortly 💯

matteo-pacini commented:
@qubitnano Please find the Darwin patch attached.

Changes on top of your implementation:
- Added Darwin deps
    -  `Metal.hpp`, a macOS Metal C++ wrapper header library
    - `apple-sdk_13` (macOS 13.0+), required by  the above
    - `spdlog`
- Removed hardcoded "/build/source/*" paths, they do not work on Darwin
    -  Rewritten `imgui'` and `stormlib'` using `applyPatches`, so they end up in the nix store, and we don't need to specify absolute paths
- Simplified `postInstallPhase`(few things have changed since 8.x.x)

![Screenshot 2025-04-04 at 10 04 54](https://github.com/user-attachments/assets/f7fcd4f8-7780-4b14-9faf-26e932c37d41)
---

```patch
diff --git a/pkgs/by-name/sh/shipwright/package.nix b/pkgs/by-name/sh/shipwright/package.nix
index e9ffa1bcacab..572091d7e448 100644
--- a/pkgs/by-name/sh/shipwright/package.nix
+++ b/pkgs/by-name/sh/shipwright/package.nix
@@ -1,4 +1,5 @@
 {
+  apple-sdk_13,
   stdenv,
   cmake,
   lsb-release,
@@ -6,7 +7,6 @@
   lib,
   fetchFromGitHub,
   fetchurl,
-  fetchpatch,
   copyDesktopItems,
   makeDesktopItem,
   python3,
@@ -27,6 +27,8 @@
   tinyxml-2,
   spdlog,
   writeTextFile,
+  fixDarwinDylibNames,
+  applyPatches,
   shipwright,
 }:
 
@@ -41,11 +43,16 @@ let
     hash = "sha256-rXC4akz9BaKzr/C2CryZC6RGk6+fGVG7RsQryUFUUk0=";
   };
 
-  imgui' = fetchFromGitHub {
-    owner = "ocornut";
-    repo = "imgui";
-    tag = "v1.91.6-docking";
-    hash = "sha256-28wyzzwXE02W5vbEdRCw2iOF8ONkb3M3Al8XlYBvz1A=";
+  imgui' = applyPatches {
+    src = fetchFromGitHub {
+      owner = "ocornut";
+      repo = "imgui";
+      tag = "v1.91.6-docking";
+      hash = "sha256-28wyzzwXE02W5vbEdRCw2iOF8ONkb3M3Al8XlYBvz1A=";
+    };
+    patches = [
+      "${shipwright.src}/libultraship/cmake/dependencies/patches/imgui-fixes-and-config.patch"
+    ];
   };
 
   libgfxd = fetchFromGitHub {
@@ -76,19 +83,31 @@ let
     hash = "sha256-xUsVponmofMsdeLsI6+kQuPg436JS3PBl00IZ5sg3Vw=";
   };
 
-  stormlib' = fetchFromGitHub {
-    owner = "ladislav-zezula";
-    repo = "StormLib";
-    tag = "v9.25";
-    hash = "sha256-HTi2FKzKCbRaP13XERUmHkJgw8IfKaRJvsK3+YxFFdc=";
+  stormlib' = applyPatches {
+    src = fetchFromGitHub {
+      owner = "ladislav-zezula";
+      repo = "StormLib";
+      tag = "v9.25";
+      hash = "sha256-HTi2FKzKCbRaP13XERUmHkJgw8IfKaRJvsK3+YxFFdc=";
+    };
+    patches = [
+      "${shipwright.src}/libultraship/cmake/dependencies/patches/stormlib-optimizations.patch"
+    ];
   };
 
   thread_pool = fetchFromGitHub {
     owner = "bshoshany";
     repo = "thread-pool";
-    rev = "v4.1.0";
+    tag = "v4.1.0";
     hash = "sha256-zhRFEmPYNFLqQCfvdAaG5VBNle9Qm8FepIIIrT9sh88=";
   };
+
+  metalcpp = fetchFromGitHub {
+    owner = "briaguya-ai";
+    repo = "single-header-metal-cpp";
+    tag = "macOS13_iOS16";
+    hash = "sha256-CSYIpmq478bla2xoPL/cGYKIWAeiORxyFFZr0+ixd7I";
+  };
 in
 stdenv.mkDerivation (finalAttrs: {
   pname = "shipwright";
@@ -123,6 +142,7 @@ stdenv.mkDerivation (finalAttrs: {
     ++ lib.optionals stdenv.hostPlatform.isDarwin [
       libicns
       darwin.sigtool
+      fixDarwinDylibNames
     ];
 
   buildInputs =
@@ -132,28 +152,37 @@ stdenv.mkDerivation (finalAttrs: {
       SDL2
       SDL2_net
       libpng
-    ]
-    ++ lib.optionals stdenv.hostPlatform.isLinux [
-      libpulseaudio
-      zenity
       libzip
       nlohmann_json
       tinyxml-2
       spdlog
+    ]
+    ++ lib.optionals stdenv.hostPlatform.isLinux [
+      libpulseaudio
+      zenity
+    ]
+    ++ lib.optionals stdenv.hostPlatform.isDarwin [
+      # Metal.hpp requires macOS 13.x min.
+      apple-sdk_13
     ];
 
-  cmakeFlags = [
-    (lib.cmakeBool "BUILD_REMOTE_CONTROL" true)
-    (lib.cmakeBool "NON_PORTABLE" true)
-    (lib.cmakeFeature "CMAKE_INSTALL_PREFIX" "${placeholder "out"}/lib")
-    (lib.cmakeFeature "FETCHCONTENT_SOURCE_DIR_IMGUI" "/build/source/build/_deps/imgui-src")
-    (lib.cmakeFeature "FETCHCONTENT_SOURCE_DIR_LIBGFXD" "${libgfxd}")
-    (lib.cmakeFeature "FETCHCONTENT_SOURCE_DIR_PRISM" "${prism}")
-    (lib.cmakeFeature "FETCHCONTENT_SOURCE_DIR_STORMLIB" "/build/source/build/_deps/stormlib-src")
-    (lib.cmakeFeature "FETCHCONTENT_SOURCE_DIR_THREADPOOL" "${thread_pool}")
-  ];
+  cmakeFlags =
+    [
+      (lib.cmakeBool "BUILD_REMOTE_CONTROL" true)
+      (lib.cmakeBool "NON_PORTABLE" true)
+      (lib.cmakeFeature "CMAKE_INSTALL_PREFIX" "${placeholder "out"}/lib")
+      (lib.cmakeFeature "FETCHCONTENT_SOURCE_DIR_IMGUI" "${imgui'}")
+      (lib.cmakeFeature "FETCHCONTENT_SOURCE_DIR_LIBGFXD" "${libgfxd}")
+      (lib.cmakeFeature "FETCHCONTENT_SOURCE_DIR_PRISM" "${prism}")
+      (lib.cmakeFeature "FETCHCONTENT_SOURCE_DIR_STORMLIB" "${stormlib'}")
+      (lib.cmakeFeature "FETCHCONTENT_SOURCE_DIR_THREADPOOL" "${thread_pool}")
+    ]
+    ++ lib.optionals stdenv.hostPlatform.isDarwin [
+      (lib.cmakeFeature "FETCHCONTENT_SOURCE_DIR_METALCPP" "${metalcpp}")
+      (lib.cmakeFeature "FETCHCONTENT_SOURCE_DIR_SPDLOG" "${spdlog}")
+    ];
 
-  env.NIX_CFLAGS_COMPILE = lib.optionalString stdenv.hostPlatform.isDarwin "-Wno-int-conversion -Wno-implicit-int";
+  env.NIX_CFLAGS_COMPILE = lib.optionalString stdenv.hostPlatform.isDarwin "-Wno-int-conversion -Wno-implicit-int -Wno-elaborated-enum-base";
 
   dontAddPrefix = true;
 
@@ -165,24 +194,7 @@ stdenv.mkDerivation (finalAttrs: {
     cp ${stb'} ./stb/${stb'.name}
     cp ${stb_impl} ./stb/${stb_impl.name}
     substituteInPlace libultraship/cmake/dependencies/common.cmake \
-      --replace-fail "\''${STB_DIR}" "/build/source/stb"
-  '';
-
-  postPatch = ''
-    # use the soh patched deps when building
-
-    mkdir -p ./build/_deps/imgui-src
-    cp -R --no-preserve=mode,ownership ${imgui'}/* ./build/_deps/imgui-src
-    pushd ./build/_deps/imgui-src
-    patch -p1 < ${shipwright.src}/libultraship/cmake/dependencies/patches/imgui-fixes-and-config.patch
-    popd
-
-    mkdir -p ./build/_deps/stormlib-src
-    cp -R --no-preserve=mode,ownership ${stormlib'}/* ./build/_deps/stormlib-src
-    pushd ./build/_deps/stormlib-src
-    patch -p1 < ${shipwright.src}/libultraship/cmake/dependencies/patches/stormlib-optimizations.patch
-    popd
-
+      --replace-fail "\''${STB_DIR}" "$(readlink -f ./stb)"
   '';
 
   postBuild = ''
@@ -215,11 +227,6 @@ stdenv.mkDerivation (finalAttrs: {
 
       mv $out/MacOS $out/Applications/soh.app/Contents/MacOS
 
-      # Wrapper
-      cp $src/soh/macosx/soh-macos.sh.in $out/Applications/soh.app/Contents/MacOS/soh
-      chmod +x $out/Applications/soh.app/Contents/MacOS/soh
-      patchShebangs $out/Applications/soh.app/Contents/MacOS/soh
-
       # "lib" contains all resources that are in "Resources" in the official bundle.
       # We move them to the right place and symlink them back to $out/lib,
       # as that's where the game expects them.
@@ -231,19 +238,8 @@ stdenv.mkDerivation (finalAttrs: {
       # Copy icons
       cp -r ../build/macosx/soh.icns $out/Applications/soh.app/Contents/Resources/soh.icns
 
-      # Fix executable
-      install_name_tool -change @executable_path/../Frameworks/libSDL2-2.0.0.dylib \
-                        ${SDL2}/lib/libSDL2-2.0.0.dylib \
-                        $out/Applications/soh.app/Contents/Resources/soh-macos
-      install_name_tool -change @executable_path/../Frameworks/libGLEW.2.2.0.dylib \
-                        ${glew}/lib/libGLEW.2.2.0.dylib \
-                        $out/Applications/soh.app/Contents/Resources/soh-macos
-      install_name_tool -change @executable_path/../Frameworks/libpng16.16.dylib \
-                        ${libpng}/lib/libpng16.16.dylib \
-                        $out/Applications/soh.app/Contents/Resources/soh-macos
-
       # Codesign (ad-hoc)
-      codesign -f -s - $out/Applications/soh.app/Contents/Resources/soh-macos
+      codesign -f -s - $out/Applications/soh.app/Contents/MacOS/soh
     '';
 
   fixupPhase = lib.optionalString stdenv.hostPlatform.isLinux ''

```

qubitnano committed:
shipwright: update darwin dependencies and use applyPatches

Added Darwin deps
    Metal.hpp, a macOS Metal C++ wrapper header library
    apple-sdk_13 (macOS 13.0+), required by the above
    spdlog

Removed hardcoded "/build/source/*" paths, they do not work on Darwin
    Rewritten imgui' and stormlib' using applyPatches, so they end up in
the nix store, and we don't need to specify absolute paths

Simplified postInstallPhase(few things have changed since 8.x.x)

Author: Matteo Pacini <m@matteopacini.me>

qubitnano commented:
@matteo-pacini Thank you I've applied your diff and credited you

matteo-pacini reviewed (approved):
LGTM, thanks @qubitnano !

## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 395867`

---
### `x86_64-darwin`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>shipwright</li>
  </ul>
</details>

---
### `aarch64-darwin`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>shipwright</li>
  </ul>
</details>

WheelsForReals commented:
I have been running this build locally and noticed that the version information in the About section is missing (Settings -> General, right column). The official AppImage displays `Blair Alfa (9.0.0)` here.

I see some CMake stuff that tries to use git commands to pull branch/commit/tag info, but I'm not quite knowledgeable enough with CMake nor nixpkgs builds to know what's causing it but I thought I'd bring it to your attention in case it is something that you want to fix.

qubitnano committed:
shipwright: include git build details

These git commands can't run in the sandbox, so populate them ourselves

qubitnano commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 395867`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>shipwright</li>
  </ul>
</details>

matteo-pacini reviewed (approved):
LGTM - version info shows correctly now, well done!

![Screenshot 2025-04-18 at 10 55 48](https://github.com/user-attachments/assets/b4d22636-2649-42bf-b644-5ab1be8cbd5f)

## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 395867`

---
### `x86_64-darwin`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>shipwright</li>
  </ul>
</details>

---
### `aarch64-darwin`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>shipwright</li>
  </ul>
</details>

Sirius902 commented:
@qubitnano Was playing around with this and noticed this wouldn't work for dev builds (non-tagged commits) so I went ahead and fixed it, here's the patch. With this you should be able to override the source to be the latest `develop` commit for example.

Tested against the current develop commit HarbourMasters/Shipwright@52a3058926c0f28f5cc077f0b6e89ee5728eb672 on both Linux and macOS.

### Changes
- Fallback to empty string for `GIT_COMMIT_TAG` (this is the behavior of the CMake as well).
https://github.com/HarbourMasters/Shipwright/blob/1b6dc13491460462f3a04e56b8fe1a28ada5d2a3/CMakeLists.txt#L68-L70
- Grab `--port-ver` from the CMake cache instead of passing `finalAttrs.version`. For tagged releases that works fine but the game expects the `--port-ver` to match the CMake project version even for dev builds so failing to do this will result in the game failing to boot since it thinks there's a version mismatch between the executable and the OTR in the derivation.

```diff
diff --git a/pkgs/by-name/sh/shipwright/package.nix b/pkgs/by-name/sh/shipwright/package.nix
index 91bc93f90f6b..baba8dab5f9c 100644
--- a/pkgs/by-name/sh/shipwright/package.nix
+++ b/pkgs/by-name/sh/shipwright/package.nix
@@ -124,7 +124,7 @@ stdenv.mkDerivation (finalAttrs: {
       cd $out
       git branch --show-current > GIT_BRANCH
       git rev-parse --short=7 HEAD > GIT_COMMIT_HASH
-      git describe --tags --abbrev=0 --exact-match HEAD > GIT_COMMIT_TAG
+      (git describe --tags --abbrev=0 --exact-match HEAD 2>/dev/null || echo "") > GIT_COMMIT_TAG
       rm -rf .git
     '';
   };
@@ -213,10 +213,11 @@ stdenv.mkDerivation (finalAttrs: {
   '';
 
   postBuild = ''
+    port_ver=$(grep CMAKE_PROJECT_VERSION: "$PWD/CMakeCache.txt" | cut -d= -f2)
     cp ${gamecontrollerdb}/gamecontrollerdb.txt gamecontrollerdb.txt
     mv ../libultraship/src/graphic/Fast3D/shaders ../soh/assets/custom
     pushd ../OTRExporter
-    python3 ./extract_assets.py -z ../build/ZAPD/ZAPD.out --norom --xml-root ../soh/assets/xml --custom-assets-path ../soh/assets/custom --custom-otr-file soh.otr --port-ver ${finalAttrs.version}
+    python3 ./extract_assets.py -z ../build/ZAPD/ZAPD.out --norom --xml-root ../soh/assets/xml --custom-assets-path ../soh/assets/custom --custom-otr-file soh.otr --port-ver $port_ver
     popd
   '';
 

```

qubitnano committed:
shipwright: fix dev builds

Fallback to empty string for GIT_COMMIT_TAG (this is the behavior of the
CMake as well).
https://github.com/HarbourMasters/Shipwright/blob/1b6dc13491460462f3a04e56b8fe1a28ada5d2a3/CMakeLists.txt#L68-L70

Grab --port-ver from the CMake cache instead of passing
finalAttrs.version. For tagged releases that works fine but the game
expects the --port-ver to match the CMake project version even for dev
builds so failing to do this will result in the game failing to boot
since it thinks there's a version mismatch between the executable and
the OTR in the derivation.

Author: Sirius902

qubitnano commented:
@Sirius902  Thank you, I've applied your patch. I've just noticed this when rebasing most of these changes onto my 2ship builds with the develop branch.

qubitnano commented:
> @qubitnano Hmm one other thing I happened to think of is I wonder if we should be using the same approach for replacing `@CMAKE_PROJECT_VERSION@` in the `Info.plist` here as is done for `--port-ver` now.
> 
> https://github.com/NixOS/nixpkgs/blob/10d91c0f2fb83586c91579ee5981c14f174b237e/pkgs/by-name/sh/shipwright/package.nix#L242
> 
> I don't think it's necessary since it still seems to run fine on darwin regardless so I'm wondering if this is just a purely informational version number and isn't used at runtime like the `--port-ver`. Not sure what's best here.
> 
> EDIT: Checked and it looks like it's just showing up when you right click on the app under "Version" if you click "Get Info" so I think it's not a big deal.

Not that familiar with macOS but yes that file seems to just be for packaging metadata such as when searching with Finder. It's probably useful to keep the git version if one is overriding. But then again [Apple also states these fields  should be 3 period separated integers](https://developer.apple.com/documentation/bundleresources/information-property-list/cfbundleshortversionstring). @matteo-pacini what do you suggest, is this good to go?

matteo-pacini commented:
@qubitnano Good catch. `CFBundleShortVersionString` is just metadata, and Apple expects a plain SemVer value there (e.g. `9.0.5`). In theory, macOS tools could read it for update checks, but those tools aren’t part of a typical Nix workflow. So I’m fine with leaving it as‑is.

matteo-pacini commented:
@qubitnano did a final rebuild as a final check, all looking good on my end:

## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 395867`

---
### `x86_64-darwin`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>shipwright</li>
  </ul>
</details>

---
### `aarch64-darwin`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>shipwright</li>
  </ul>
</details>

Happy to merge if all comments / issues are resolved

keenanweaver reviewed (approved):
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 395867`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>shipwright</li>
  </ul>
</details>
